### PR TITLE
WD-10729 - update contact us forms

### PR DESCRIPTION
**Pull request by lizzochek**
## Done

Updated multiple contact forms. Please see more details in the Jira ticket.  

## QA

- demos:
- https://ubuntu-com-13918.demos.haus/support/contact-us?product=support-overview
- https://ubuntu-com-13918.demos.haus/contact-us/form?product=pro
- https://ubuntu-com-13918.demos.haus/security/esm#get-in-touch
- https://ubuntu-com-13918.demos.haus/pricing/pro#get-in-touch
- https://ubuntu-com-13918.demos.haus/ceph/consulting#get-in-touch
- https://ubuntu-com-13918.demos.haus/core/contact-us
- https://ubuntu-com-13918.demos.haus/server/contact-us
- https://ubuntu-com-13918.demos.haus/openstack/contact-us
- https://ubuntu-com-13918.demos.haus/kernel/real-time/contact-us
- https://ubuntu-com-13918.demos.haus/aws/contact-us
- https://ubuntu-com-13918.demos.haus/aws#get-in-touch
- https://ubuntu-com-13918.demos.haus/landscape/pricing#get-in-touch
- https://ubuntu-com-13918.demos.haus/landscape/features#get-in-touch
- https://ubuntu-com-13918.demos.haus/desktop/organisations#get-in-touch
- View the site on demo
- Run through the following [QA steps](https://discourse.canonical.com/t/qa-steps/152)
- Check that the forms are correct according to copy docs

## Issue / Card
[WD-10729](https://warthogs.atlassian.net/browse/WD-10729)

[WD-10729]: https://warthogs.atlassian.net/browse/WD-10729?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by juanruitina** [comment]
- [ ] Please correctly capitalise GitHub in every instance.
- [ ] https://ubuntu-com-13918.demos.haus/core/contact-us: update [copy doc](https://docs.google.com/document/d/1sS0KAfhTyDxsOtooZ2-4cUEy-vqvtSpSpzgrX-TbfJM/edit) metatag
- [ ] https://ubuntu-com-13918.demos.haus/aws/contact-us: update [copy doc](https://docs.google.com/document/d/1Qa1FfWQVadxxn9zqEf3V5MLYyHoBLwyTJD7fICnfX_w/edit) metatag

A bigger issue. I don't think we should use the 50/50 split when a modal form has several sections. Unlike brochure pages where users might skim, users will likely fill forms sequentially. I'd favour keeping the current simple layout. Currently they are paginated: I don't oppose this, at least "About you" section ([see it at the end of the copy docs](https://docs.google.com/document/d/1jy_dKnQu6Nh_wsk3S7GxYfJZZsvEGT89LA7_UMpqfWs/edit)). @lyubomir-popov, what do you think? I'll reach out to Jared too to see what he thinks.

These are the modals:

- https://ubuntu-com-13918.demos.haus/security/esm#get-in-touch
- https://ubuntu-com-13918.demos.haus/pricing/pro#get-in-touch
- https://ubuntu-com-13918.demos.haus/ceph/consulting#get-in-touch
- https://ubuntu-com-13918.demos.haus/aws#get-in-touch
- https://ubuntu-com-13918.demos.haus/landscape/pricing#get-in-touch
- https://ubuntu-com-13918.demos.haus/landscape/features#get-in-touch
- https://ubuntu-com-13918.demos.haus/desktop/organisations#get-in-touch

Also checking with Jared:

- [ ] https://ubuntu-com-13918.demos.haus/kernel/real-time/contact-us: top section looks weird, suggesting change in [copy doc](https://docs.google.com/document/d/1974sNirEiujc0I3OSBs_n9bb6vz4YorK7at8jK-Ljj8/edit#heading=h.a44alhvalg6q)
- [ ] https://ubuntu-com-13918.demos.haus/desktop/organisations#get-in-touch: checking the content for this too, [copy doc](https://docs.google.com/document/d/1G5bHrvTcQ3gsYiGQHD94xB8BrIEBW04f7Sn_ZClTuJY/edit) doesn't seem update.

**Comment by lyubomir-popov** [comment]
@juanruitina I think the problem is that it isn't implemented correctly yet, lines missing, double indents, so it looks confusing. 

@lizzochek Can we plase have the double grid indent removed, so all text lines up on the left and right, and then ca nwe please have the horizontal rules above each h2? I think then we'd be in a better place to evaluate whether the design is working or not.

**Comment by lizzochek** [comment]
@lyubomir-popov @juanruitina you can see an updated version in this PR: https://github.com/canonical/canonical.com/pull/1270 
I corrected according to @lyubomir-popov comments in there

**Comment by lyubomir-popov** [comment]
also, can we please do h4s and change the title of the popup to an h3 - lots of long questions, so we have to adapt
image

**Comment by juanruitina** [comment]
@lyubomir-popov Do we have any research on how 50/50 performs on forms? As mentioned above, I'd expect users to fill a form sequentially, and with the 50/50 split they would end up zigzagging. I think it's OK to have two columns if left is an intro to the form and right the fields, but I'm not convinced about splitting fields like that.

**Comment by lizzochek** [comment]
@lyubomir-popov updated in https://github.com/canonical/canonical.com/pull/1270

**Comment by juanruitina** [comment]
Got confirmation from Jared on the two items I mentioned at the end:

- [ ] https://ubuntu-com-13918.demos.haus/kernel/real-time/contact-us: top section looks weird, suggesting change in [copy doc](https://docs.google.com/document/d/1974sNirEiujc0I3OSBs_n9bb6vz4YorK7at8jK-Ljj8/edit#heading=h.a44alhvalg6q)
- [ ] https://ubuntu-com-13918.demos.haus/desktop/organisations#get-in-touch: [copy doc](https://docs.google.com/document/d/1G5bHrvTcQ3gsYiGQHD94xB8BrIEBW04f7Sn_ZClTuJY/edit) was not updated, please have a look.

**Comment by carkod** [comment]
Can you resolve the conflict, so I can have a clear view of the `thank-you` page code?

File: static/sass/styles.scss [hunk 1]
```diff
@@ -821,6 +822,16 @@ td.p-accordion {
...
 
+#contact-success {
+  display: none;
+  position: absolute;
+}
+
+#contact-success:target {
+  display: block;
+  z-index: 10;
+}
```
Comment (carkod) [inline static/sass/styles.scss:833]: Are these used anywhere? I can't find any code that uses `#contact-success`

File: static/js/src/contact-forms.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+function validateCheckbox(event, fieldsetId) {
+    const checkboxes = Array.from(document.getElementById(fieldsetId).querySelectorAll("input[class='p-checkbox__input']"));
+    if (event.currentTarget.checked) {
+        checkboxes[0].removeAttribute("required");
+    }
+}
+
+function getRadioItemValue(fieldset) {
+    const selectedRadio = fieldset.querySelector("input[name='how-many-machines-do-you-have']:checked")
```
Comment (carkod) [inline static/js/src/contact-forms.js:9]: I don't know if you wrote this piece of code, but it seems uncessary? Duplication of line `42`:

```
const inputs = howManyMachinesFieldset?.querySelectorAll("input[name='how-many-machines-do-you-have']")
```

File: static/sass/styles.scss (hunk 1 above)
Comment (carkod) [reply static/sass/styles.scss:833]: Also, in the case we are actually using these styles, can you move them to `_pattern_modals.scss`, would make more sense if they are used across all form modals to be there?

File: templates/kernel/real-time/contact-us.html [hunk 3]
```diff
@@ -327,3 +661,5 @@ <h3 class="p-heading--4">Your comments</h3>
   </section>
 
 {% endblock %}
+
+<script src="/static/js/src/contact-forms.js"></script>
```
Comment (carkod) [inline templates/kernel/real-time/contact-us.html:665]: Ideally, we should use [versioned_static](https://github.com/search?q=repo%3Acanonical%2Fubuntu.com%20versioned_static&type=code), will accelerate loading time and allow content-cache to cache this for website failures.

File: templates/security/thank-you.html [hunk 4]
```diff
@@ -17,7 +17,7 @@
 
 {% elif product == 'esm' %}
 
-  {% with thanks_context="about Expanded Security Maintenance" %}{% include "shared/_thank_you.html" %}{% endwith %}
+  {% with tthanks_context="contacting our team." thanks_message="We will be in touch shortly." %}{% include "shared/_thank_you.html" %}{% endwith %}
```
Comment (carkod) [inline templates/security/thank-you.html:20]: Typo "tthanks_contenxt"

File: templates/shared/_client-contact-us-form.html [hunk 5]
```diff
@@ -1,13 +1,16 @@
 <section class="p-strip is-deep">
   <div class="row">
     <div class="col-8">
-      <h1>{{h1}}</h1>
-      <p>{{intro_text}}</p>
+      <h1>{{ h1 }}</h1>
```
Comment (carkod) [inline templates/shared/_client-contact-us-form.html:4]: Also ideally, we shouldn't name things the same as HTML tags or programming language keywords, so we don't mistakengly override things. In this case it should be fine, it's HTML, but it's a good habit to not name it this way.

File: templates/shared/_cloud-contact-us-form.html [hunk 6]
```diff
@@ -1,43 +1,391 @@
 <section class="p-strip">
   <div class="row">
     <div class="col-8">
-      <h1>{{h1}}</h1>
-      <p class="p-heading--4">{{intro_text}}</p>
+      <h1>{{ h1 }}</h1>
```
Comment (carkod) [inline templates/shared/_cloud-contact-us-form.html:4]: Same here, ideaily variable names shouldn't match with HTML tags

File: templates/shared/_cloud-contact-us-form.html [hunk 7]
```diff
@@ -1,43 +1,391 @@
...
+              </label>
+            </div>
+          </div>
+        </fieldset>
+        <fieldset class="u-no-margin--bottom" id="how-many-machines">
+          <legend class="u-off-screen">*How many devices?</legend>
+
+          <h2 class="p-heading--3">
+            *How many
+            {% if howManyVM %}
```
Comment (carkod) [inline templates/shared/_cloud-contact-us-form.html:162]: This condition is wrong, the value assigned is a string `"true"`, so `"false"` would be truthy. Either replace this line to `howManyVM === "true"` or the value you pass in the include should be a boolean:
```
    ....
    formid="3473",
    lpId="",
    returnURL="/aws/thank-you",
    howManyVM=True %}
```

Notice the boolean howManyVM `True`

File: templates/templates/base.html [hunk 8]
```diff
@@ -160,6 +160,18 @@
                                     </div>
                                 </div>
                             </div>
+                            <div id="contact-success" class="p-strip u-no-padding--top">
```
Comment (carkod) [inline templates/templates/base.html:163]: Ok I see the styles are used here. Ignore the comment in `styles.scss`

**Comment by carkod** [comment]
There's a bug in https://ubuntu-com-13918.demos.haus/security/esm#get-in-touch (and I guess every page that uses that component):

When I click on one of the options (required) of "What kind of device are you using?*", it doesn't recognize it as selected.

Screenshot 2024-06-17 at 14 35 08

Also another issue is that there are fields that are required, but then there is no asterisk, such as:
Screenshot 2024-06-17 at 14 42 44

Can you add the [is-required classname](https://vanillaframework.io/docs/base/forms#required) for those that have the `required` attribute?

Thanks

File: templates/kernel/real-time/contact-us.html [hunk 9]
```diff
@@ -197,16 +171,376 @@ <h3 class="p-heading--4">How are you currently meeting your latency requirements
...
+                <input required id="firstName" name="firstName" maxlength="255" type="text" />
+              </li>
+              <li class="p-list__item">
+                <label for="lastName">Last name:</label>
+                <input required id="lastName" name="lastName" maxlength="255" type="text" />
+              </li>
+              <li class="p-list__item">
+                <label for="company">Company:</label>
+                <input required id="company" name="company" maxlength="255" type="text" />
+              </li>
```
Comment (carkod) [inline templates/kernel/real-time/contact-us.html:525]: Missing asterisk (required field) for some of these, please check https://docs.google.com/document/d/1974sNirEiujc0I3OSBs_n9bb6vz4YorK7at8jK-Ljj8/edit

File: templates/kernel/real-time/contact-us.html [hunk 10]
```diff
@@ -197,16 +171,376 @@ <h3 class="p-heading--4">How are you currently meeting your latency requirements
...
+                <input required
+                       id="email"
+                       name="email"
+                       maxlength="255"
+                       type="email"
+                       pattern="^[^ ]+@[^ ]+\.[a-z]{2,26}$" />
+              </li>
+              <li class="p-list__item">
+                <label for="phone">Mobile/cell phone number:</label>
+                <input required id="phone" name="phone" maxlength="255" type="tel" />
```
Comment (carkod) [inline templates/kernel/real-time/contact-us.html:541]: This is also required, maybe we should have the `is-required` class?

**Comment by juanruitina** [comment]
Some more feedback. There are quite a few issues with these forms accessibility wise. This kind of markup is very problematic:

```html
      <div class="row--50-50 p-section--shallow u-no-padding--left u-no-padding--right">
        <div class="col">
          <legend class="u-off-screen">Tell us about your project and team</legend>
          <h4>Tell us about your project</h4>
        </div>
        <div class="col">
          <textarea id="about-your-project" rows="3"></textarea>
        </div>
      </div>
```

[Every input should have a label associated](https://www.w3.org/WAI/tutorials/forms/labels/#associating-labels-explicitly). The H4 should be a label element with H4 styling and refer to the input field.

Also, we should not use that off-screen legend attribute as we do. [Legend elements are used to label groups of fields in fieldsets](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/legend), but we are abusing them for no good reason. Screen readers would read both legend and the h4s, making them redundant. Legend would only make sense here as far as I can see for "About us", that could be the legend for a fieldset with all the fields that we have on the right column.

Something cleaner would look like this:

```html
      <div class="row--50-50 p-section--shallow u-no-padding--left u-no-padding--right">
        <div class="col">
          <label class="p-heading--4" for="about-your-project">Tell us about your project</label>
        </div>
        <div class="col">
          <textarea name="about-your-project" id="about-your-project" rows="3"></textarea>
        </div>
      </div>
```

(Still inquiring about the two-column layout, will get back to you today)

File: static/sass/styles.scss (hunk 1 above)
Comment (lizzochek) [reply static/sass/styles.scss:833]: This is a confirmation banner for landscape form. The banners themselves are a part of base_index, so all banner styles are saved in styles, so I added it there too

File: static/js/src/contact-forms.js (hunk 2 above)
Comment (lizzochek) [reply static/js/src/contact-forms.js:9]: They are not, first one checks for the selected values as the field is required, second selects all inputs

**Comment by lizzochek** [comment]
@carkod @juanruitina updated according to your comments

**Comment by carkod** [comment]
> * real-time

The main two bugs are still there? Please check https://github.com/canonical/ubuntu.com/pull/13918#issuecomment-2173436694

**Comment by carkod** [comment]
I've given code +1 but please fix the linting errors.

**Comment by juanruitina** [comment]
> * Please correctly capitalise GitHub in every instance.
> * [x]  https://ubuntu-com-13918.demos.haus/core/contact-us: update [copy doc](https://docs.google.com/document/d/1sS0KAfhTyDxsOtooZ2-4cUEy-vqvtSpSpzgrX-TbfJM/edit) metatag
> * [x]  https://ubuntu-com-13918.demos.haus/aws/contact-us: update [copy doc](https://docs.google.com/document/d/1Qa1FfWQVadxxn9zqEf3V5MLYyHoBLwyTJD7fICnfX_w/edit) metatag

These don't seem addressed in demos.

Following up on the "legend" fields, I see them being useful for groups of checkboxes and radio buttons (should be fieldsets). Both fieldsets and legends have some styles coming from either Vanilla or browser styles. A cleaner markup for checkbox/radio fieldsets would look like this:

```html
      <fieldset class="row--50-50 p-section--shallow u-no-padding" id="kind-of-device">
        <div class="col">
          <legend class="p-heading--4">What kind of device are you using?*</legend>
        </div>
        <div class="col">
          <div class="row u-no-padding">
            <div class="col-4 u-sv3">
              <label class="p-checkbox">
                <input required="" class="p-checkbox__input" type="checkbox" aria-labelledby="desktop/workstation" value="desktop/workstation" onclick="validateCheckbox(event, 'kind-of-device')">
                <span class="p-checkbox__label" id="desktop-workstation">Desktop/workstation</span>
              </label>
              [...]
            </div>
          </div>
        </div>
      </fieldset>
```

Border would need to be removed from the fieldset, I couldn't find a utility for that. Happy to pair to make this work, some further testing would be useful.

**Comment by eabashidze** [comment]
Hi team,
I tested the forms, and how the values sync to Marketo from them. You can see the outcomes [in this document](https://docs.google.com/spreadsheets/d/1Wjq3p5wkGjDxxMD9oqD59AZRR8nOTTTUo1AZINAUn3o/edit?usp=sharing) (let me know if you need further access). I also ran into one error, which I could not replicate and will upload the photo here:

**Comment by laszlokajtar** [comment]
@carkod there are multiple Marketo submission and sync issues. my hunch is that it has to do with how the form submission has changed with the addition of the 

onsubmit="getCustomFields(event)" attribute

**Comment by carkod** [comment]
> getCustomFields(e

What kind of problems are you having? Can you be more specific?

**Comment by laszlokajtar** [comment]
@carkod the document linked by Erekle above has the information. mainly: we are receiving only email addresses in Marketo and not the contents of the rest of the fields even though they appear in the payload. in the case of comments from lead field, the value of that doesn't appear in the payload. i suppose that Marketo submission doesn't wait for the returining of the getCustomFields function.

**Comment by eabashidze** [comment]
Hi team, I have updated [the report](https://docs.google.com/spreadsheets/d/1Wjq3p5wkGjDxxMD9oqD59AZRR8nOTTTUo1AZINAUn3o/edit?usp=sharing) with today's test results (included UTMs in the test). No forms have synced any fields this time unfortunately, only email addresses (see the sheet - marketo database Jun 20 test).

**Comment by carkod** [comment]
> Hi team, I have updated [the report](https://docs.google.com/spreadsheets/d/1Wjq3p5wkGjDxxMD9oqD59AZRR8nOTTTUo1AZINAUn3o/edit?usp=sharing) with today's test results (included UTMs in the test). No forms have synced any fields this time unfortunately, only email addresses (see the sheet - marketo database Jun 20 test).

@eabashidze Thanks. Are we having the same issues on the live site?

**Comment by lyubomir-popov** [comment]
Please change the background of all forms from white to #f3f3f3 (we have a class is-paper I believe, pls don't hardcode the hex value)

**Comment by lyubomir-popov** [comment]
This is not using the 50/50 split - either all should or not IMO.

**Comment by lyubomir-popov** [comment]
Most of the forms use the outgoing style as opposed to the rebranding, not sure if I'm missing anything?

**Comment by britneywwc** [comment]
Moving this task to a feature branch, will address the design and form submission issues there.

**Pull request merged**